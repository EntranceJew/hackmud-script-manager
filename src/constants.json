{
	"supportedExtensions": [ ".js", ".ts" ],
	"validDBMethods": [ "i", "r", "f", "u", "u1", "us", "ObjectId" ]
}
