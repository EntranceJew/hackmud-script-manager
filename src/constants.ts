export const supportedExtensions = [ `.js`, `.ts` ]
export const validDBMethods = [ `i`, `r`, `f`, `u`, `u1`, `us`, `ObjectId` ]
