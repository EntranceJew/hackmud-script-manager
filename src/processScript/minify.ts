import babelGenerator from "@babel/generator"
import { parse } from "@babel/parser"
import babelTraverse from "@babel/traverse"
import t, { Expression } from "@babel/types"
import { assert, getHackmudCharacterCount, spliceString } from "@samual/lib"
import { tokenizer as tokenize, tokTypes as tokenTypes } from "acorn"
import * as terser from "terser"

const { default: generate } = babelGenerator as any as typeof import("@babel/generator")
const { default: traverse } = babelTraverse as any as typeof import("@babel/traverse")

// TODO when there are more than 3 references to `$G`, place a `let _GLOBAL_0_ = $G` at the top and replace references with `_GLOBAL_0_`
// TODO move autocomplete stuff outside this function
// TODO allow not mangling class and function names

/**
 * @param code compiled code and/or hackmud compatible code
 * @param autocomplete the comment inserted after the function signature
 * @param uniqueID 11 a-z 0-9 characters
 */
export async function minify(code: string, autocomplete: string, uniqueID = "00000000000") {
	assert(uniqueID.match(/^\w{11}$/))

	const jsonValues: any[] = []
	let undefinedIsReferenced = false

	// remove dead code (so we don't waste chracters quine cheating strings
	// that aren't even used)
	code = (await terser.minify(code, {
		ecma: 2015,
		parse: { bare_returns: true },
		compress: {
			passes: Infinity,
			unsafe: true,
			booleans: false,
			sequences: false
		}
	})).code || ""

	let scriptBeforeJSONValueReplacement

	{
		// BUG when this script is used, the source char count is off

		const file = await parse(code)

		traverse(file, {
			MemberExpression({ node: memberExpression }) {
				if (memberExpression.computed)
					return

				assert(memberExpression.property.type == "Identifier")

				if (memberExpression.property.name == "prototype") {
					memberExpression.computed = true
					memberExpression.property = t.identifier(`_PROTOTYPE_PROPERTY_${uniqueID}_`)
				} else if (memberExpression.property.name == "__proto__") {
					memberExpression.computed = true
					memberExpression.property = t.identifier(`_PROTO_PROPERTY_${uniqueID}_`)
				}
			}
		})

		scriptBeforeJSONValueReplacement = (await terser.minify(generate(file!).code, {
			ecma: 2015,
			compress: {
				passes: Infinity,
				unsafe: true,
				unsafe_arrows: true,
				unsafe_comps: true,
				unsafe_symbols: true,
				unsafe_methods: true,
				unsafe_proto: true,
				unsafe_regexp: true,
				unsafe_undefined: true,
				sequences: false
			},
			format: { semicolons: false }
		})).code!
			.replace(new RegExp(`_PROTOTYPE_PROPERTY_${uniqueID}_`, "g"), `"prototype"`)
			.replace(new RegExp(`_PROTO_PROPERTY_${uniqueID}_`, "g"), `"__proto__"`)
	}

	let comment: string | null = null
	let hasComment = false

	{
		const file = await parse(code)
		const promises: Promise<any>[] = []

		traverse(file, {
			FunctionDeclaration(path) {
				path.traverse({
					Function(path) {
						if (path.parent.type != "CallExpression" && path.parentKey != "callee")
							path.skip()
					},

					Loop(path) {
						path.skip()
					},

					ObjectExpression(path) {
						const o: Record<string, unknown> = {}

						if (parseObjectExpression(path.node, o))
							path.replaceWith(t.identifier(`_JSON_VALUE_${jsonValues.push(o) - 1}_${uniqueID}_`))
					},

					ArrayExpression(path) {
						const o: unknown[] = []

						if (parseArrayExpression(path.node, o))
							path.replaceWith(t.identifier(`_JSON_VALUE_${jsonValues.push(o) - 1}_${uniqueID}_`))
					}
				})

				path.traverse({
					TemplateLiteral(path) {
						const templateLiteral = path.node
						let replacement: babel.Node = t.stringLiteral(templateLiteral.quasis[0].value.cooked!)

						for (let i = 0; i < templateLiteral.expressions.length; i++) {
							const expression = templateLiteral.expressions[i] as Expression
							const templateElement = templateLiteral.quasis[i + 1]

							replacement = t.binaryExpression(
								"+",
								replacement,
								expression
							)

							if (!templateElement.value.cooked)
								continue

							replacement = t.binaryExpression(
								"+",
								replacement,
								t.stringLiteral(templateElement.value.cooked!)
							)
						}

						path.replaceWith(replacement)
					},

					MemberExpression({ node: memberExpression }) {
						if (memberExpression.computed)
							return

						assert(memberExpression.property.type == "Identifier")

						if (memberExpression.property.name.length < 3)
							return

						memberExpression.computed = true
						memberExpression.property = t.stringLiteral(memberExpression.property.name)
					},

					UnaryExpression(path) {
						if (path.node.operator == "void" && path.node.argument.type == "NumericLiteral" && !path.node.argument.value) {
							path.replaceWith(t.identifier(`_UNDEFINED_${uniqueID}_`))
							undefinedIsReferenced = true
						}
					},

					NullLiteral(path) {
						let jsonValueIndex = jsonValues.indexOf(null)

						if (jsonValueIndex == -1)
							jsonValueIndex += jsonValues.push(null)

						path.replaceWith(t.identifier(`_JSON_VALUE_${jsonValueIndex}_${uniqueID}_`))
					},

					BooleanLiteral(path) {
						let jsonValueIndex = jsonValues.indexOf(path.node.value)

						if (jsonValueIndex == -1)
							jsonValueIndex += jsonValues.push(path.node.value)

						path.replaceWith(t.identifier(`_JSON_VALUE_${jsonValueIndex}_${uniqueID}_`))
					},

					NumericLiteral(path) {
						promises.push((async () => {
							if ((await minifyNumber(path.node.value)).length <= 3)
								return

							if (path.parentKey == "key" && path.parent.type == "ObjectProperty")
								path.parent.computed = true

							let jsonValueIndex = jsonValues.indexOf(path.node.value)

							if (jsonValueIndex == -1)
								jsonValueIndex += jsonValues.push(path.node.value)

							path.replaceWith(t.identifier(`_JSON_VALUE_${jsonValueIndex}_${uniqueID}_`))
						})())
					},

					StringLiteral(path) {
						if (path.node.value.includes("\u0000") || path.node.value.length < 2)
							return

						if (path.parentKey == "key" && path.parent.type == "ObjectProperty")
							path.parent.computed = true

						let jsonValueIndex = jsonValues.indexOf(path.node.value)

						if (jsonValueIndex == -1)
							jsonValueIndex += jsonValues.push(path.node.value)

						path.replaceWith(t.identifier(`_JSON_VALUE_${jsonValueIndex}_${uniqueID}_`))
					},

					ObjectProperty({ node }) {
						if (node.computed || node.key.type != "Identifier" || node.key.name.length < 4)
							return

						let jsonValueIndex = jsonValues.indexOf(node.key.name)

						if (jsonValueIndex == -1)
							jsonValueIndex += jsonValues.push(node.key.name)

						node.computed = true
						node.key = t.identifier(`_JSON_VALUE_${jsonValueIndex}_${uniqueID}_`)
					}
				})

				path.skip()
			}
		})

		await Promise.all(promises)

		const [ functionDeclaration ] = file.program.body

		assert(functionDeclaration.type == "FunctionDeclaration")

		if (jsonValues.length) {
			hasComment = true

			if (jsonValues.length == 1) {
				if (typeof jsonValues[0] == "string" && !jsonValues[0].includes("\n") && !jsonValues[0].includes("\t")) {
					const variableDeclaration = t.variableDeclaration(
						"let",
						[
							t.variableDeclarator(
								t.identifier(`_JSON_VALUE_0_${uniqueID}_`),
								t.memberExpression(
									t.taggedTemplateExpression(
										t.memberExpression(
											t.callExpression(t.identifier(`$${uniqueID}$SUBSCRIPT$scripts$quine`), []),
											t.identifier("split")
										),
										t.templateLiteral([ t.templateElement({ raw: "\t", cooked: "\t" }, true) ], [])
									),
									t.identifier(`$${uniqueID}$SPLIT_INDEX`),
									true
								)
							)
						]
					)

					if (undefinedIsReferenced)
						variableDeclaration.declarations.push(t.variableDeclarator(t.identifier(`_UNDEFINED_${uniqueID}_`)))

					functionDeclaration.body.body.unshift(variableDeclaration)

					comment = jsonValues[0]
				} else {
					const variableDeclaration = t.variableDeclaration(
						"let",
						[
							t.variableDeclarator(
								t.identifier(`_JSON_VALUE_0_${uniqueID}_`),
								t.callExpression(
									t.memberExpression(
										t.identifier("JSON"),
										t.identifier("parse")
									),
									[
										t.memberExpression(
											t.taggedTemplateExpression(
												t.memberExpression(
													t.callExpression(t.identifier(`$${uniqueID}$SUBSCRIPT$scripts$quine`), []),
													t.identifier("split")
												),
												t.templateLiteral([ t.templateElement({ raw: "\t", cooked: "\t" }, true) ], [])
											),
											t.identifier(`$${uniqueID}$SPLIT_INDEX`),
											true
										)
									]
								)
							)
						]
					)

					if (undefinedIsReferenced)
						variableDeclaration.declarations.push(t.variableDeclarator(t.identifier(`_UNDEFINED_${uniqueID}_`)))

					functionDeclaration.body.body.unshift(variableDeclaration)

					comment = JSON.stringify(jsonValues[0])
				}
			} else {
				const variableDeclaration = t.variableDeclaration(
					"let",
					[
						t.variableDeclarator(
							t.arrayPattern(jsonValues.map((_, i) => t.identifier(`_JSON_VALUE_${i}_${uniqueID}_`))),
							t.callExpression(
								t.memberExpression(
									t.identifier("JSON"),
									t.identifier("parse")
								),
								[
									t.memberExpression(
										t.taggedTemplateExpression(
											t.memberExpression(
												t.callExpression(t.identifier(`$${uniqueID}$SUBSCRIPT$scripts$quine`), []),
												t.identifier("split")
											),
											t.templateLiteral([ t.templateElement({ raw: "\t", cooked: "\t" }, true) ], [])
										),
										t.identifier(`$${uniqueID}$SPLIT_INDEX`),
										true
									)
								]
							)
						)
					]
				)

				if (undefinedIsReferenced)
					variableDeclaration.declarations.push(t.variableDeclarator(t.identifier(`_UNDEFINED_${uniqueID}_`)))

				functionDeclaration.body.body.unshift(variableDeclaration)

				comment = JSON.stringify(jsonValues)
			}
		} else if (undefinedIsReferenced) {
			functionDeclaration.body.body.unshift(
				t.variableDeclaration(
					"let",
					[ t.variableDeclarator(t.identifier(`_UNDEFINED_${uniqueID}_`)) ]
				)
			)
		}

		code = generate(file!).code
	}

	code = (await terser.minify(code, {
		ecma: 2015,
		compress: {
			passes: Infinity,
			unsafe: true,
			unsafe_arrows: true,
			unsafe_comps: true,
			unsafe_symbols: true,
			unsafe_methods: true,
			unsafe_proto: true,
			unsafe_regexp: true,
			unsafe_undefined: true,
			sequences: false
		},
		format: { semicolons: false }
	})).code || ""


	// this step affects the character count and can't be done after the count comparison
	if (comment != null) {
		code = spliceString(code, `${autocomplete ? `//${autocomplete}\n` : ""}\n//\t${comment}\t\n`, getFunctionBodyStart(code) + 1)

		for (const [ i, part ] of code.split("\t").entries()) {
			if (part == comment) {
				code = code.replace(`$${uniqueID}$SPLIT_INDEX`, await minifyNumber(i))
				break
			}
		}
	}

	// if the script has a comment, it's gonna contain `SC$scripts$quine()`
	// which is gonna eventually compile to `#fs.scripts.quine()` which contains
	// an extra character so we have to account for that
	if (getHackmudCharacterCount(scriptBeforeJSONValueReplacement) <= (getHackmudCharacterCount(code) + Number(hasComment))) {
		code = scriptBeforeJSONValueReplacement

		if (autocomplete)
			code = spliceString(code, `//${autocomplete}\n`, getFunctionBodyStart(code) + 1)
	}

	return code
}

export default minify

function parseObjectExpression(node: babel.types.ObjectExpression, o: Record<string, unknown>) {
	if (!node.properties.length)
		return false

	for (const property of node.properties) {
		if (property.type != "ObjectProperty" || property.computed)
			return false

		assert(property.key.type == "Identifier" || property.key.type == "NumericLiteral" || property.key.type == "StringLiteral")

		if (property.value.type == "ArrayExpression") {
			const childArray: unknown[] = []

			if (parseArrayExpression(property.value, childArray))
				o[property.key.type == "Identifier" ? property.key.name : property.key.value] = childArray
			else
				return false
		} else if (property.value.type == "ObjectExpression") {
			const childObject: Record<string, unknown> = {}

			if (parseObjectExpression(property.value, childObject))
				o[property.key.type == "Identifier" ? property.key.name : property.key.value] = childObject
			else
				return false
		} else if (property.value.type == "NullLiteral")
			o[property.key.type == "Identifier" ? property.key.name : property.key.value] = null
		else if (property.value.type == "BooleanLiteral" || property.value.type == "NumericLiteral" || property.value.type == "StringLiteral")
			o[property.key.type == "Identifier" ? property.key.name : property.key.value] = property.value.value
		else
			return false
	}

	return true
}

function parseArrayExpression(node: babel.types.ArrayExpression, o: unknown[]) {
	if (!node.elements.length)
		return false

	for (const element of node.elements) {
		if (!element)
			return false

		if (element.type == "ArrayExpression") {
			const childArray: unknown[] = []

			if (parseArrayExpression(element, childArray))
				childArray.push(childArray)
			else
				return false
		} else if (element.type == "ObjectExpression") {
			const childObject: Record<string, unknown> = {}

			if (parseObjectExpression(element, childObject))
				o.push(childObject)
			else
				return false
		} else if (element.type == "NullLiteral")
			o.push(null)
		else if (element.type == "BooleanLiteral" || element.type == "NumericLiteral" || element.type == "StringLiteral")
			o.push(element.value)
		else
			return false
	}

	return true
}

async function minifyNumber(number: number) {
	return (await terser.minify(`$(${number})`, { ecma: 2015 })).code!.match(/\$\((.+)\)/)![1]
}

function getFunctionBodyStart(code: string) {
	const tokens = tokenize(code, { ecmaVersion: 2015 })

	tokens.getToken() // function
	tokens.getToken() // name
	tokens.getToken() // (

	let nests = 1

	while (nests) {
		const token = tokens.getToken()

		if (token.type == tokenTypes.parenL)
			nests++
		else if (token.type == tokenTypes.parenR)
			nests--
	}

	return tokens.getToken().start // {
}
