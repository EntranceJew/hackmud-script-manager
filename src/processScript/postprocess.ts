export function postprocess(code: string, seclevel: number, uniqueID: string) {
	return code
		.replace(/^function\s*\w+\(/, `function(`)
		.replace(new RegExp(`\\$${uniqueID}\\$\\\\\\\\0\\$SC_DOLLAR\\$`, `g`), `SC\\$`)
		.replace(new RegExp(`\\$${uniqueID}\\$\\\\\\\\0\\$DB_DOLLAR\\$`, `g`), `DB\\$`)
		.replace(new RegExp(`\\$${uniqueID}\\$\\\\\\\\0\\$D\\$`, `g`), `__D_\\S`)
		.replace(new RegExp(`\\$${uniqueID}\\$\\\\\\\\0\\$FMCL\\$`, `g`), `__FMCL\\_`)
		.replace(new RegExp(`\\$${uniqueID}\\$\\\\\\\\0\\$G\\$`, `g`), `__G\\_`)
		.replace(new RegExp(`\\$${uniqueID}\\$SUBSCRIPT\\$(\\w+)\\$(\\w+)`, `g`), `#${`nlmhf`[seclevel]}s.$1.$2`)
		.replace(new RegExp(`\\$${uniqueID}\\$DEBUG`, `g`), `#D`)
		.replace(new RegExp(`\\$${uniqueID}\\$FMCL`, `g`), `#FMCL`)
		.replace(new RegExp(`\\$${uniqueID}\\$GLOBAL`, `g`), `#G`)
		.replace(new RegExp(`\\$${uniqueID}\\$DB\\$(\\w+)`, `g`), `#db.$1`)
		.replace(new RegExp(`\\$${uniqueID}\\$SLASH_SLASH\\$`, `g`), `/\\/`)
}

export default postprocess
