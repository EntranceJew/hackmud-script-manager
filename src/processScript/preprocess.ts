import babelGenerator from "@babel/generator"
import { parse } from "@babel/parser"
import babelTraverse, { NodePath } from "@babel/traverse"
import t, { Program } from "@babel/types"
import { assert, spliceString } from "@samual/lib"

const { default: traverse } = babelTraverse as any as typeof import("@babel/traverse")
const { default: generate } = babelGenerator as any as typeof import("@babel/generator")

export type PreprocessOptions = {
	/** 11 a-z 0-9 characters */
	uniqueID: string
}

/**
 * @param code source code for preprocessing
 * @param options {@link PreprocessOptions details}
 */
export function preprocess(code: string, { uniqueID = `00000000000` }: Partial<PreprocessOptions> = {}) {
	assert(/^\w{11}$/.test(uniqueID))

	// TODO rename variables that trips this
	if (/(?:SC|DB)\$/.test(code))
		throw new Error(`SC$ and DB$ are protected and cannot appear in a script`)

	const sourceCode = code
	let lengthBefore

	do {
		lengthBefore = code.length

		code = code
			.replace(/^\s+/, ``)
			.replace(/^\/\/.*/, ``)
			.replace(/^\/\*[\s\S]*?\*\//, ``)
	} while (code.length != lengthBefore)

	code = code.replace(/^function\s*\(/, `export default function (`)

	let file

	while (true) {
		let error

		try {
			file = parse(code, {
				plugins: [
					`typescript`,
					[ `decorators`, { decoratorsBeforeExport: true } ],
					`doExpressions`,
					`functionBind`,
					`functionSent`,
					`partialApplication`,
					[ `pipelineOperator`, { proposal: `hack`, topicToken: `%` } ],
					`throwExpressions`,
					[ `recordAndTuple`, { syntaxType: `hash` } ],
					`classProperties`,
					`classPrivateProperties`,
					`classPrivateMethods`,
					`logicalAssignment`,
					`numericSeparator`,
					`nullishCoalescingOperator`,
					`optionalChaining`,
					`optionalCatchBinding`,
					`objectRestSpread`
				],
				sourceType: `module`
			})

			break
		} catch (error_) {
			assert(error_ instanceof SyntaxError)

			error = error_ as SyntaxError & {
				pos: number
				code: string
				reasonCode: string
			}
		}

		if (error.code != `BABEL_PARSER_SYNTAX_ERROR` || error.reasonCode != `PrivateInExpectedIn`) {
			console.log((/.+/.exec(code.slice(error.pos)))?.[0])

			throw error
		}

		const codeSlice = code.slice(error.pos)
		let match

		if ((match = /^#[0-4fhmln]s\.scripts\.quine\(\)/.exec(codeSlice)))
			code = spliceString(code, JSON.stringify(sourceCode), error.pos, match[0]!.length)
		else if ((match = /^#[0-4fhmln]?s\./.exec(codeSlice)))
			code = spliceString(code, `$`, error.pos, 1)
		else if ((match = /^#D[^\w$]/.exec(codeSlice)))
			code = spliceString(code, `$`, error.pos, 1)
		else if ((match = /^#FMCL/.exec(codeSlice)))
			code = spliceString(code, `$${uniqueID}$FMCL`, error.pos, match[0]!.length)
		else if ((match = /^#G/.exec(codeSlice)))
			code = spliceString(code, `$${uniqueID}$GLOBAL`, error.pos, match[0]!.length)
		else if ((match = /^#db\./.exec(codeSlice)))
			code = spliceString(code, `$`, error.pos, 1)
		else
			throw error
	}

	let program!: NodePath<Program>

	traverse(file, {
		Program(path) {
			program = path
			path.skip()
		}
	})

	const needRecord = program.scope.hasGlobal(`Record`)
	const needTuple = program.scope.hasGlobal(`Tuple`)

	if (needRecord || needTuple) {
		file.program.body.unshift(t.importDeclaration(
			needRecord
				? (needTuple
					? [
						t.importSpecifier(t.identifier(`Record`), t.identifier(`Record`)),
						t.importSpecifier(t.identifier(`Tuple`), t.identifier(`Tuple`))
					]
					: [ t.importSpecifier(t.identifier(`Record`), t.identifier(`Record`)) ]
				) : [ t.importSpecifier(t.identifier(`Tuple`), t.identifier(`Tuple`)) ],
			t.stringLiteral(`@bloomberg/record-tuple-polyfill`)
		))
	}

	if (program.scope.hasGlobal(`Proxy`)) {
		file.program.body.unshift(t.importDeclaration([
			t.importDefaultSpecifier(t.identifier(`Proxy`))
		], t.stringLiteral(`proxy-polyfill/src/proxy.js`)))
	}

	if (program.node.body.length == 1 && program.node.body[0]!.type == `FunctionDeclaration`)
		return { code: `export default ${generate(file).code}` }

	return { code: generate(file).code }
}

export default preprocess
