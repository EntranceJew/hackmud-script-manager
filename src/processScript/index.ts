import babelGenerator from "@babel/generator"
import { parse } from "@babel/parser"
import babelPluginProposalClassProperties from "@babel/plugin-proposal-class-properties"
import babelPluginProposalClassStaticBlock from "@babel/plugin-proposal-class-static-block"
import babelPluginProposalDecorators from "@babel/plugin-proposal-decorators"
import babelPluginProposalDoExpressions from "@babel/plugin-proposal-do-expressions"
import babelPluginProposalFunctionBind from "@babel/plugin-proposal-function-bind"
import babelPluginProposalFunctionSent from "@babel/plugin-proposal-function-sent"
import babelPluginProposalJSONStrings from "@babel/plugin-proposal-json-strings"
import babelPluginProposalLogicalAssignmentOperators from "@babel/plugin-proposal-logical-assignment-operators"
import babelPluginProposalNullishCoalescingOperator from "@babel/plugin-proposal-nullish-coalescing-operator"
import babelPluginProposalNumericSeparator from "@babel/plugin-proposal-numeric-separator"
import babelPluginProposalObjectRestSpread from "@babel/plugin-proposal-object-rest-spread"
import babelPluginProposalOptionalCatchBinding from "@babel/plugin-proposal-optional-catch-binding"
import babelPluginProposalOptionalChaining from "@babel/plugin-proposal-optional-chaining"
import babelPluginProposalPartialApplication from "@babel/plugin-proposal-partial-application"
import babelPluginProposalPipelineOperator from "@babel/plugin-proposal-pipeline-operator"
import babelPluginProposalPrivatePropertyInObject from "@babel/plugin-proposal-private-property-in-object"
import babelPluginProposalRecordAndTuple from "@babel/plugin-proposal-record-and-tuple"
import babelPluginProposalThrowExpressions from "@babel/plugin-proposal-throw-expressions"
import babelPluginTransformExponentiationOperator from "@babel/plugin-transform-exponentiation-operator"
import babelPluginTransformTypescript from "@babel/plugin-transform-typescript"
import babelTraverse from "@babel/traverse"
import t from "@babel/types"
import rollupPluginBabel_ from "@rollup/plugin-babel"
import rollupPluginCommonJS from "@rollup/plugin-commonjs"
import rollupPluginJSON from "@rollup/plugin-json"
import rollupPluginNodeResolve from "@rollup/plugin-node-resolve"
import { assert, countHackmudCharacters } from "@samual/lib"
import { resolve as resolvePath } from "path"
import { performance } from "perf_hooks"
import prettier from "prettier"
import { rollup } from "rollup"
import { supportedExtensions as extensions } from "../constants.json"
import minify from "./minify"
import postprocess from "./postprocess"
import preprocess from "./preprocess"
import transform from "./transform"

const { default: rollupPluginBabel } = rollupPluginBabel_ as any as typeof import("@rollup/plugin-babel")
const { format } = prettier
const { default: generate } = babelGenerator as any as typeof import("@babel/generator")
const { default: traverse } = babelTraverse as any as typeof import("@babel/traverse")

export { minify } from "./minify"
export { postprocess } from "./postprocess"
export { preprocess } from "./preprocess"
export { transform as compile } from "./transform"

export type ProcessOptions = {
	/** whether to minify the given code */
	minify: boolean

	/** 11 a-z 0-9 characters */
	uniqueID: string

	/** the user going to be hosting this script (or set to `true` if not yet known) */
	scriptUser: string | true

	/** the name of this script (or set to `true` if not yet known) */
	scriptName: string | true

	filePath: string

	/** whether to mangle function and class names (defaults to `false`) */
	mangleNames: boolean
}

/**
 * Minifies a given script
 *
 * @param code JavaScript or TypeScript code
 * @param options {@link ProcessOptions details}
 */
export async function processScript(
	code: string,
	{
		minify: shouldMinify = true,
		uniqueID = Math.floor(Math.random() * (2 ** 52)).toString(36).padStart(11, `0`),
		scriptUser = `UNKNOWN`,
		scriptName = `UNKNOWN`,
		filePath,
		mangleNames = false
	}: Partial<ProcessOptions> = {}
): Promise<{
	srcLength: number
	script: string
	warnings: { message: string, line: number }[]
	timeTook: number
}> {
	assert(/^\w{11}$/.exec(uniqueID))

	const time = performance.now()
	const sourceCode = code
	let autocomplete
	let statedSeclevel

	// TODO do seclevel detection and verification per module

	const autocompleteMatch = /^function\s*\(.+\/\/(?<autocomplete>.+)/.exec(code)

	if (autocompleteMatch) {
		code = `export default ${code}`;
		({ autocomplete } = autocompleteMatch.groups!)
	} else {
		for (const line of code.split(`\n`)) {
			const comment = /^\s*\/\/(?<commentContent>.+)/.exec(line)

			if (!comment)
				break

			const commentContent = comment.groups!.commentContent!.trim()

			if (commentContent.startsWith(`@autocomplete `))
				autocomplete = commentContent.slice(14).trimStart()
			else if (commentContent.startsWith(`@seclevel `)) {
				const seclevelString = commentContent.slice(10).trimStart().toLowerCase()

				switch (seclevelString) {
					case `fullsec`:
					case `full`:
					case `fs`:
					case `4s`:
					case `f`:
					case `4`: {
						statedSeclevel = 4
					} break

					case `highsec`:
					case `high`:
					case `hs`:
					case `3s`:
					case `h`:
					case `3`: {
						statedSeclevel = 3
					} break

					case `midsec`:
					case `mid`:
					case `ms`:
					case `2s`:
					case `m`:
					case `2`: {
						statedSeclevel = 2
					} break

					case `lowsec`:
					case `low`:
					case `ls`:
					case `1s`:
					case `l`:
					case `1`: {
						statedSeclevel = 1
					} break

					case `nullsec`:
					case `null`:
					case `ns`:
					case `0s`:
					case `n`:
					case `0`: {
						statedSeclevel = 0
					} break

					default:
						// TODO turn into warninig when I get round to those
						throw new Error(`unrecognised seclevel "${seclevelString}"`)
				}
			}
		}
	}

	assert(/^\w{11}$/.exec(uniqueID))

	const filePathResolved = filePath
		? resolvePath(filePath)
		: `${uniqueID}.ts`

	const bundle = await rollup({
		input: filePathResolved,
		plugins: [
			{
				name: `hackmud-script-manager`,
				transform: code => preprocess(code, { uniqueID }).code
			},
			rollupPluginBabel({
				babelHelpers: `bundled`,
				plugins: [
					[ babelPluginTransformTypescript.default, { allowDeclareFields: true, optimizeConstEnums: true } ],
					[ babelPluginProposalDecorators.default, { decoratorsBeforeExport: true } ],
					[ babelPluginProposalDoExpressions.default ],
					[ babelPluginProposalFunctionBind.default ],
					[ babelPluginProposalFunctionSent.default ],
					[ babelPluginProposalPartialApplication.default ],
					[ babelPluginProposalPipelineOperator.default, { proposal: `hack`, topicToken: `%` } ],
					[ babelPluginProposalThrowExpressions.default ],
					[ babelPluginProposalRecordAndTuple.default, { syntaxType: `hash`, importPolyfill: true } ],
					[ babelPluginProposalClassProperties.default ],
					[ babelPluginProposalClassStaticBlock.default ],
					[ babelPluginProposalPrivatePropertyInObject.default ],
					[ babelPluginProposalLogicalAssignmentOperators.default ],
					[ babelPluginProposalNumericSeparator.default ],
					[ babelPluginProposalNullishCoalescingOperator.default ],
					[ babelPluginProposalOptionalChaining.default ],
					[ babelPluginProposalOptionalCatchBinding.default ],
					[ babelPluginProposalJSONStrings.default ],
					[ babelPluginProposalObjectRestSpread.default ],
					[ babelPluginTransformExponentiationOperator.default ]
				],
				configFile: false,
				extensions
			}),
			rollupPluginCommonJS(),
			rollupPluginNodeResolve({ extensions }),
			rollupPluginJSON()
		],
		treeshake: { moduleSideEffects: false }
	})

	const seclevelNames = [ `NULLSEC`, `LOWSEC`, `MIDSEC`, `HIGHSEC`, `FULLSEC` ]

	code = (await bundle.generate({})).output[0].code

	const { file, seclevel } = transform(parse(code, { sourceType: `module` }), sourceCode, { uniqueID, scriptUser, scriptName })

	if (statedSeclevel != undefined && seclevel < statedSeclevel)
		// TODO replace with a warning and build script anyway
		throw new Error(`detected seclevel ${seclevelNames[seclevel]} is lower than stated seclevel ${seclevelNames[statedSeclevel]}`)

	code = generate(file).code

	// TODO fix incorrect source length again

	// the typescript inserts semicolons where they weren't already so we take
	// all semicolons out of the count and add the number of semicolons in the
	// source to make character count fairer
	const sourceLength = countHackmudCharacters(code.replace(/^function\s*\w+\(/, `function(`))
		// - (code.match(/;/g)?.length || 0)
		// + semicolons
		// + (code.match(/SC\$[a-zA-Z_][a-zA-Z0-9_]*\$[a-zA-Z_][a-zA-Z0-9_]*\(/g)?.length ?? 0)
		// + (code.match(/DB\$/g)?.length ?? 0)

	if (shouldMinify)
		code = await minify(file, autocomplete, { uniqueID, mangleNames })
	else {
		traverse(file, {
			MemberExpression({ node: memberExpression }) {
				if (memberExpression.computed)
					return

				assert(memberExpression.property.type == `Identifier`)

				if (memberExpression.property.name == `prototype`) {
					memberExpression.computed = true
					memberExpression.property = t.stringLiteral(`prototype`)
				} else if (memberExpression.property.name == `__proto__`) {
					memberExpression.computed = true
					memberExpression.property = t.stringLiteral(`__proto__`)
				}
			}
		})

		code = format(generate(file).code, {
			parser: `babel`,
			arrowParens: `avoid`,
			semi: false,
			trailingComma: `none`
		})
	}

	code = postprocess(code, seclevel, uniqueID)

	return {
		srcLength: sourceLength,
		script: code,
		warnings: [],
		timeTook: performance.now() - time
	}
}

export default processScript
