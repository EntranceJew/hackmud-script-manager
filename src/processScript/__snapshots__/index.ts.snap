// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`programmatic api works 1`] = `
{
  "script": "function(){#4s.scripts.trust()}",
  "warnings": [],
}
`;
